go/apps/mixnet/socks5.go

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2015, Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mixnet
+
+import (
+	"errors"
+	"net"
+	"strconv"
+)
+
+// Version of the SOCKS protocol expected to be implemented by the client.
+const SocksVersion = 0x05
+
+// SocksListener implements the net.Listener interface as a SOCKS server. This
+// program partially implements the server role in version 5 of the SOCKS
+// protocol specified in RFC 1928. In particular, it only supports TCP clients
+// with no authentication who request CONNECT to IPv4 addresses; neither BIND
+// nor UDP ASSOCIATE are supported.
+type SocksListener struct {
+	net.Listener
+	network string // Network protocol for proxying, e.g. "tcp".
+}
+
+// SocksConn implements the net.Conn interface and contains a destination
+// network and address for the proxy.
+type SocksConn struct {
+	net.Conn
+	network, addr string // Destination network and address.
+}
+
+// Get destination address of connection.
+func (c *SocksConn) GetAddr() string {
+	return c.addr
+}
+
+// Accept exposes the SOCKS5 protocol to connecting client. Return the
+// connection and the requested destination address.
+func (l *SocksListener) Accept() (net.Conn, error) {
+	c, err := l.Listener.Accept()
+	if err != nil {
+		return nil, err
+	}
+
+	// First, wait for greeting from client containing the SOCKS version and
+	// requested methods.
+	msg := make([]byte, MaxMsgBytes)
+	reply := make([]byte, MaxMsgBytes)
+	if _, err = c.Read(msg); err != nil {
```

Comment:
Also get the length and reduce the slice to the number of bytes received. You should use this to make sure you sent enough bytes, and you should add a test where you connect with too few bytes to make sure this works. As it currently stands, this will panic if the client sends, e.g., only 1 byte.